ci/requirements.txt

```diff
@@ -0,0 +1,10 @@
+Pillow
```

Comment:
Missing torch requirements.
torchtext, torchvision etc.. ?

Reply:
The torch package installation command is different for CPU and GPU and should not be part of requirements.txt. Also this is CI specific and the torch packages are installed through the codebuild docker files.

Reply:
what would happen if I tried installing with just instructions based on ``requirements.txt`` ? I would fail to run examples etc. correct?

Reply:
This reqirements.txt is specific to CI build and code_coverage where the torch related packages are taken care by docker files. However, for manually running torchserve_sanity.sh script, the torch packages will need to be installed separately.

Reply:
Updated documentation to install torchserve pre-requisites before running sanity script.

Reply:
Why do we need a requirements.txt then?
Why are you saying this txt file is specific to CI build or code coverage ? What prevents us from using it everywhere?

The main point of using requirements.txt is to have one central place where we list all the dependencies. It doesn't matter what we have today in ci scripts, sanity scripts etc.
Let's list all the deps in one single place (requirements.txt) and install it from different scripts. This will be used by pip, conda etc.

The idea is not to miss any installations that we are experiencing.

Reply:
@mycpuorg : The only reason to keep torch related packages away from the requirements.txt file is the fact that `PyTorch`'s official documentation provides different set of installation commands for torch packages based on the environment like Linux/Mac, CPU/GPU, Pip/Conda etc. Following are some combination which we generally use :

|Environment  | Installation command  |
| ------------- | ------------- |
| Linux + Pip + Cuda | pip install torch torchvision
| Linux + Pip | pip install torch==1.5.0+cpu torchvision==0.6.0+cpu -f https://download.pytorch.org/whl/torch_stable.html
| Linux + Conda | conda install pytorch torchvision cpuonly -c pytorch
| Linux + Conda + Cuda | conda install pytorch torchvision cudatoolkit=10.2 -c pytorch
| Mac + Pip | pip install torch torchvision
| Mac + Conda | conda install pytorch torchvision -c pytorch

Reference : https://pytorch.org/get-started/locally/

In case of CI build the torch packages are installed as part of the docker image used for CodeBuild and this requirements.txt was introduced initially to freeze all the python packages being used at the time of CI build.

Also, following packages are required only while executing test suites or python linting on the backend source code and model-archiver source code and not required for installing `TorchServe` from source :

```
mock
pytest
pylint
pytest-mock
pytest-cov
```

Note that the pytest suite is currently not executed as a part of build process.

For this we can split the requirements.txt into multiple parts like this :
```
|-- requirements
|   |-- common.txt
|   |-- dev.txt
|   `-- prod.txt
`-- requirements.txt
```

common.txt
```
## contains common packages like
future
psutil
wheel
requests
```

dev.txt
```
## replicates the common.txt and adds dev specific packages like :
mock
pytest
pylint
pytest-mock
pytest-cov
```

prod.txt
```
## replicates common.txt
-r common.txt
```

requirements.txt
```
## replicates prod.txt
-r prod.txt
```

Please let us know your thoughts and how should we handle the different installation commands provided by PyTorch for torch packages.

Reply:
Thanks. I like the idea of modularizing the installation. But it's not clear what the expectation is for installing torch (specific) packages. The whole issue that we are trying to solve is to list and install the dep packages from one central place. Is something breaking when we do this?

Reply:
Also, please describe what your proposal is and how the installation commands would look afterwards (forget CI, Dockerfiles, sanity scripts). Just list down the commands that an end user would be typing into the CLI after these requirements.txt changes are merged.

Reply:
@mycpuorg :

The idea here is to use 2 different requirements.txt for installing the environment specific packages for production and development envs. 

For dev environment or CI builds we will use

`pip install -r requirements/dev_requirements.txt`

which will install all the common packages along with the UT specific python packages.

For production environment we will use :

`pip install requirements.txt` or `pip install requirements/production.txt`

This will install all the common packages and anything specific to production environment.

Regarding the torch related packages :

The installation works fine with the following pip install command which is being currently used in the installation scripts :

`pip install torch torchtext torchvision`

However, given the different commands recommended by PyTorch for different environments, this will required a bit deeper testing on different environments using different models to be sure we can use one common command for all the environments.

Reply:
@harshbafna @mycpuorg What is the final resolution of this discussion? What is being proposed now and what has to be done as a future step?

Reply:
@chauhang @mycpuorg 

I have tested the install_from_source_ubuntu script and verified all the default handlers on EC2 with CPU/GPU and Pip/Conda combination. Based on this testing we should be good to go ahead with the above mentioned approach of modularized requirements.txt which is also a best practice for keeping the production environment independent of development related.

Please refer the attached excel sheet for detailed report.
[Torch Package Installation Testing.xlsx](https://github.com/pytorch/serve/files/4634431/Torch.Package.Installation.Testing.xlsx)